bundle/jsky.app.ot/src/main/scala/edu/gemini/catalog/ui/image/BackgroundImageLoader.scala

```diff
@@ -0,0 +1,220 @@
+package edu.gemini.catalog.ui.image
+
+import java.beans.{PropertyChangeEvent, PropertyChangeListener}
+import java.nio.file.Path
+import java.time.Instant
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.{ExecutorService, Executors, ThreadFactory}
+
+import edu.gemini.catalog.image._
+import edu.gemini.shared.util.immutable.ScalaConverters._
+
+import scala.collection.JavaConverters._
+import edu.gemini.pot.sp._
+import edu.gemini.spModel.core.{Coordinates, Wavelength}
+import jsky.app.ot.tpe.{ImageCatalogPanel, TpeContext, TpeImageWidget, TpeManager}
+import jsky.image.gui.ImageLoadingException
+
+import scalaz._
+import Scalaz._
+import scala.swing.Swing
+import scalaz.concurrent.Task
+
+/**
+  * Describes a requested image for an observation and wavelength
+  */
+case class TargetImageRequest(key: SPNodeKey, coordinates: Coordinates, obsWavelength: Option[Wavelength])
+
+object TargetImageRequest {
+  /** @group Typeclass Instances */
+  implicit val equal: Equal[TargetImageRequest] = Equal.equalA[TargetImageRequest]
+}
+
+/**
+  * This interface can be used to listen when the image is being loaded and update the UI
+  */
+trait ImageLoadingListener {
+  def downloadStarts(): Task[Unit]
+  def downloadCompletes(): Task[Unit]
+  def downloadError(): Task[Unit]
+}
+
+/**
+  * Listens for program changes and download images as required. It listens
+  * for program changes and requests downloading images if they are not already
+  * presents.
+  * It also updates the UI as needed
+  */
+object BackgroundImageLoader {
+  private val taskUnit = Task.now(())
+
+  private def imageDownloadsThreadFactory(priority: Int) = new ThreadFactory {
+    private val threadNumber: AtomicInteger = new AtomicInteger(1)
+    private val defaultThreadFactory = Executors.defaultThreadFactory()
+
+    override def newThread(r: Runnable): Thread = {
+      val name = s"Background Image Downloads - ${threadNumber.getAndIncrement()} - priority: $priority"
+      defaultThreadFactory.newThread(r) <| {_.setDaemon(true)} <| {_.setName(name)} <| {_.setPriority(priority)}
```

Comment:
`<|` is deprecated but we use it extensively in the OCS, particularly in Swing. I personally like it and would rather make our own `unsafeTap` once we move to scalaz 7.3